src/main/java/org/truffleruby/language/globals/GlobalVariableStorage.java

```diff
@@ -26,26 +28,35 @@
 
     private volatile Object value;
 
+    private final Object defaultValue;
     private final DynamicObject getter;
     private final DynamicObject setter;
+    private final DynamicObject isDefined;
 
-    GlobalVariableStorage(Object value, DynamicObject getter, DynamicObject setter) {
-        this.value = value;
+    GlobalVariableStorage(Object value, DynamicObject getter, DynamicObject setter, DynamicObject isDefined) {
+        this.defaultValue = value;
+        this.value = UNSET_VALUE;
```

Comment:
It would make sense to rename `value` to defaultValue here and adapt the callers to just give `nil()` for it.
Then `GlobalVariables.put` would do a setValueInternal() call to set the initial value.
Might be worth splitting in two constructor for with/without hooks.